sycl/source/handler.cpp

```diff
@@ -895,11 +895,18 @@ event handler::finalize() {
 #endif
   }
 
+  bool DiscardEvent = !impl->MEventNeeded && Queue &&
+                      Queue->supportsDiscardingPiEvents() &&
+                      CommandGroup->getRequirements().size() == 0;
+
   detail::EventImplPtr Event = detail::Scheduler::getInstance().addCG(
-      std::move(CommandGroup), Queue->shared_from_this(), impl->MEventNeeded);
+      std::move(CommandGroup), Queue->shared_from_this(), !DiscardEvent);
 
 #ifdef __INTEL_PREVIEW_BREAKING_CHANGES
-  MLastEvent = Event;
+  // For preview mode, handler.finalize() is expected to return nullptr
+  // if the event is discarded.
```

Comment:
This should be moved/copied to `handler.hpp`, IMO.

Reply:
Move to handler.hpp. Yes, we can assert that in queue, inside parseEvent(). Done.